File: RoadOfCoding/Tree/owner_solutions/v1_589.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        dfs(root,result);
        return result;
    }
    public void dfs(Node node, ArrayList<Integer> result){
        if(node == null)
            return;
        result.add(node.val);
        for(int index = 0;index<node.children.size();index++){
            dfs(node.children.get(index),result);
        }
    }
}

File: RoadOfCoding/Tree/owner_solutions/589. N-ary Tree Preorder Traversal/v1_589.js
/**
 * // Definition for a _Node.
 * function _Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {_Node|null} root
 * @return {number[]}
 */
var preorder = function(root) {
    let index = 0;
    let result = [];
    let dfs = function (node){
      if(node==null){
          return;
      }
      result.push(node.val);
      for(let idx = 0;idx<node.children.length;idx++){
          dfs(node.children[idx]);
      }
    }
      dfs(root);
      
      return result;
  };